articles/RefReturnScope.dd

```diff
@@ -0,0 +1,207 @@
+}
+---
+
+$(P I.e. if a pointer variable is set to be the address of a local variable, or to the contents of
+scope pointer, then that pointer variable is automatically set to be a scope pointer.
+The compiler is pretty good at inferring scope, thus relieving the programmer of adding
+a lot of annotations.)
+
+$(P Note that as `scope` is a storage class, not a type constructor, it is not possible to specify
+a scope pointer to a scope pointer. It surprisingly turns out to not be necessary to support that.)
+
+
+$(H3 The Role of `return scope`)
+
+$(P Consider the following:)
+
+---
+void f()
+{
+    int i;
+    *process(&i) = 4;
+}
+
+int* process(scope int* p) { return p; }
+---
+
+$(P This is perfectly legitimate code, there is no pointer escaping bug. But it won't compile.
+Without `scope` on the parameter `p`, the call to `process(&i)` would be disallowed.
+But with `scope` on `p`, the `return p;` is disallowed.)
+
+$(P The solution is adding a `return` annotation:)
+
+---
+int* process(return scope int* p) { return p; }
+---
+
+$(P which allows the scope pointer value to be returned by the function.)
+
+$(P If a function returns `void`, `return scope` also allows returning the scope value
+through the first parameter:)
+
+---
+void mun(ref int* v, return scope int* p)
+{
+    v = p; // ok
+}
+---
+
+$(P And that's it for pointers. Remember that pointer scoping is concerned with the value
+of the pointer variable.)
+
+
+$(H3 The Role of `ref`)
+
+$(P A `ref` is a reference to a value, a fancy way of representing a pointer to a value.
+It is distinguished from a pointer:)
+
+$(OL
+$(LI by not allowing arithmetic on the address)
+$(LI whenever the `ref` variable is used an automatic dereference is performed)
+$(LI a ref cannot escape from a function)
+)
+
+---
+ref int fin(ref int i)
+{
+    return i; // error, cannot return ref variable i by ref
+}
+---
+
+
+$(H3 The role of `return ref`)
+
+$(P But it will be allowed if `return` is applied:)
+
+---
+ref int fin(return ref int i)
+{
+    return i; // ok
+}
+---
+
+
+$(H3 The role of `ref scope`)
+
+$(P The storage classes `ref` and `scope` together are orthogonal, they do not affect each other.
+`ref` refers to the address of the variable, `scope` refers to the contents of the variable.)
+
+
+$(H3 The role of `return ref scope`)
+
+$(P The `return` here applies to the `ref`, not the `scope`.)
+
+$(P Let's try tricking the compiler:)
+
+---
+ref int* fin(return ref scope int* p) { return p; } // ok
+
+int* tricky()
+{
+    int i;
+    int* p = &i; // p is now inferred to be scope
+    auto q = fin(p);  // q now contains the address of i, and so scope is also inferred
+    return q;  // error: scope variable `q` may not be returned
+}
+---
+
+$(P Curses! Foiled again!)
+
+$(P The operational idea here is, while compiling @safe code, it is not be possible to escape
+a scoped value, no matter how twisty the code is.)
+
+)
+
+Macros:
+        TITLE=Coralling Wild Pointers With `ref return scope`
```

Comment:
```suggestion
        TITLE=Coralling Wild Pointers With $D(ref return scope)
```

Reply:
The backticks should do the same thing.

Reply:
I know they do in general in DDoc, but I looked at the rendered HTML and
`` `ref return scope` `` appeared verbatim in the page title. This is possibly a bug.

Reply:
I found an existing issue: https://issues.dlang.org/show_bug.cgi?id=18703

Reply:
This is an intentional part of the design: the backticks are not processed in macros so it is possible to define macros that give access to literal backticks.